relay/temp-pallets/pallet-staking-fixer/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Temporary pallet to fix corrupted staking ledgers.
+//!
+//! This single-purpose pallet exposes an extrinsic which works as a proxy to
+//! `Staking::restore_ledger`. There are three differences between the extrinsic exposed by this
+//! pallet and the exposed by the Staking pallet, namely:
+//!
+//! - Access control: any signed account may call `Call::restore_ledger`. The call to
+//! `Staking::restore_ledger` from this pallet will be performed with Origin::Root. It is safe to
+//! expose the call to the sign origin since *only* corrupted ledgers can be restored. If the stash
+//! is not corrupted, the call is a no-op (modulo paid fees).
+//! - Default inputs: `Call::restore_ledger` exposes less inputs than the proxied
+//! `Staking::restore_ledger` call, providing sensible defaults that are enough to restore all
+//! corrupted ledgers.
+//! - Extra safety check: if the resulting restored ledger from `Staking::restore_ledger` has a
+//! larger total staked than the stash's free balance, the ledger is unstaked.
+//!
+//! Note: this pallet is temporary and can be removed when all the ledgers in Polkadot and Kusama
+//! are restored.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::Currency;
+use frame_system::ensure_signed;
+use pallet_staking::WeightInfo;
+use sp_staking::StakingAccount;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::{pallet_prelude::*, RawOrigin};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_staking::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Ledger of `stash` has been recovered.
+		Restored { stash: T::AccountId },
+		/// Ledger of `stash` has been recovered. The resulting recoving ended up in unbonding and
+		/// the ledger to unstake.
+		RestoredUnstaked { stash: T::AccountId },
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Restores the state of a staking ledger that is in an inconsistent state.
+		///
+		/// This call functions as a proxy to `Staking::restore_ledger`, but exposes it to signed
+		/// origins. In addition, this call checks if, after the ledger has been successfully
+		/// restored, the ledger locks are higher or equal than the stash's free balance. If not, it
+		/// forces the unstake of the ledger.
+		///
+		/// Safety note: Only ledgers associated `stash` that are corrupted will be mutated. Thus it
```

Comment:
So you are saying that the original extrinsic was needlessly restrictive, and it is generally okay to let anyone call these?

Reply:
The `fn restore_ledger` will fail with `CannotRestoreLedger` if the ledger state does not fall into one of the corrupted ledger cases (basically, gated by [`fn inspect_bond_state`](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/pallet/impls.rs#L96C16-L96C34). The idea of requiring staking admin origin was to make it explicit through a referendum which ledgers were being restored (and for an extra layer of safety).

Reply:
To add more safety (i.e. not relying solely on [`fn inspect_bond_state`](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/pallet/impls.rs#L96C16-L96C34) to ensure that the ledger should be mutated and recovered), we add also a whitelisting mechanism so we can control which stashes need to be recovered in the configs based on the current on-chain state.